File: src/engine/camera/camera.ts
import { mat4, quat, vec3, type Mat4, type Quat, type Vec3 } from "wgpu-matrix";
import type { CameraSettings } from "./camera-settings";
import type { CameraController } from "./camera-controller";

// From https://iolite-engine.com/blog_posts/reverse_z_cheatsheet
const reverseZ = mat4.create(
  1,
  0,
  0,
  0, //
  0,
  1,
  0,
  0, //
  0,
  0,
  -1,
  1, //
  0,
  0,
  0,
  1 //
);

// Or maybe I should use https://docs.rs/glam/latest/src/glam/f32/sse2/mat4.rs.html#969-982

export class Camera {
  private _view: Mat4;
  private _projection: Mat4;
  constructor(
    public position: Vec3,
    public orientation: Quat,
    aspect: number,
    settings: CameraSettings
  ) {
    this._projection = mat4.mul(
      mat4.perspective(settings.fov, aspect, settings.zNear, settings.zFar),
      reverseZ
    );
    this._view = calculateViewMatrix(position, orientation);
  }

  update(controller: CameraController) {
    this._view = calculateViewMatrix(
      controller.getPosition(),
      controller.getOrientation()
    );
  }

  updateAspect(aspect: number) {
    // Ported from the Rust code
    this._projection[0] = -this._projection[5] / aspect;
  }

  get view(): Readonly<Mat4> {
    return this._view;
  }

  get projection(): Readonly<Mat4> {
    return this._projection;
  }

  static get forward(): Vec3 {
    return vec3.create(0, 0, -1);
  }
  static get up(): Vec3 {
    return vec3.create(0, 1, 0);
  }
  static get right(): Vec3 {
    return vec3.create(1, 0, 0);
  }
}

function calculateViewMatrix(position: Vec3, orientation: Quat): Mat4 {
  const cameraDirection = vec3.transformQuat(Camera.forward, orientation);
  const target = vec3.add(position, cameraDirection);
  return mat4.lookAt(position, target, Camera.up);
}

File: parametric-renderer-core/desktop/src/camera/camera.rs
use glam::Quat;
use glamour::{Angle, Matrix4, Point3, Vector3};

use super::{camera_controller::CameraController, camera_settings::CameraSettings};

#[derive(Debug)]
pub struct Camera {
    pub position: Point3,
    pub orientation: Quat,
    pub settings: CameraSettings,

    view: Matrix4<f32>,
    proj: Matrix4<f32>,
}

impl Camera {
    pub fn new(aspect_ratio: f32, settings: CameraSettings) -> Self {
        let position = Point3::ZERO;
        let orientation = Quat::IDENTITY;

        let proj =
            calculate_projection(aspect_ratio, settings.fov, settings.z_near, settings.z_far);

        let view = calculate_view(position, orientation);

        Self {
            position,
            orientation,
            settings,
            proj,
            view,
        }
    }

    /// Positions the camera
    pub fn view_matrix(&self) -> Matrix4<f32> {
        self.view
    }

    pub fn projection_matrix(&self) -> Matrix4<f32> {
        self.proj
    }

    pub fn update_camera(&mut self, controller: &impl CameraController) {
        self.position = controller.position();
        self.orientation = controller.orientation();

        self.view = calculate_view(self.position, self.orientation);
    }

    pub fn update_aspect_ratio(&mut self, aspect_ratio: f32) {
        // See https://docs.rs/glam/0.27.0/src/glam/f32/sse2/mat4.rs.html#969-982
        self.proj.as_cols_mut()[0][0] = -self.proj.as_cols()[1][1] / aspect_ratio;
    }

    /// in world-space
    pub const fn forward() -> Vector3 {
        Vector3::new(0.0, 0.0, -1.0)
    }

    /// in world-space
    pub const fn right() -> Vector3 {
        Vector3::new(1.0, 0.0, 0.0)
    }

    /// in world-space
    pub const fn up() -> Vector3 {
        Vector3::new(0.0, 1.0, 0.0)
    }
}

/// fov is expected to be in radians
fn calculate_projection(aspect_ratio: f32, fov: Angle, near: f32, _far: f32) -> Matrix4<f32> {
    Matrix4::perspective_infinite_reverse_rh(fov, aspect_ratio, near)
}

fn calculate_view(position: Point3, orientation: Quat) -> Matrix4<f32> {
    let cam_direction = orientation * Camera::forward();
    let target = position + cam_direction;

    Matrix4::look_at_rh(position, target, Camera::up())
}
